source/simulator/introspection.cc

```diff
@@ -32,166 +32,157 @@ namespace aspect
 {
   namespace internal
   {
-
     /**
-     * Return pair with @p n_components and a filled ComponentIndices structure.
+     * Return a filled ComponentIndices structure.
      */
     template <int dim>
-    std::pair<unsigned int, typename Introspection<dim>::ComponentIndices>
-    setup_component_indices (const unsigned int n_compositional_fields)
+    typename Introspection<dim>::ComponentIndices
+    setup_component_indices (dealii::FEVariableSystem<dim> &ib)
     {
       typename Introspection<dim>::ComponentIndices ci;
-      unsigned int comp = 0;
-
       for (unsigned int i=0; i<dim; ++i)
-        ci.velocities[i] = comp++;
-      ci.pressure = comp++;
-      ci.temperature = comp++;
+        ci.velocities[i] = ib.variable("velocity").first_component_index+i;
+
+      ci.pressure = ib.variable("pressure").first_component_index;
+      ci.temperature = ib.variable("temperature").first_component_index;
+      unsigned int n_compositional_fields = ib.variable("compositions").n_components();
       for (unsigned int i=0; i<n_compositional_fields; ++i)
-        ci.compositional_fields.push_back(comp++);
-      return std::make_pair(comp, ci);
+        ci.compositional_fields.push_back(ib.variable("compositions").first_component_index+i);
```

Comment:
On 03/03/2016 05:08 PM, Timo Heister wrote:
>
> 1. keep |introspection.component_mask.velocity| and always require that there
> is a variable "velocity" (this is the current state)
> 2. keep |introspection.component_mask.velocity| but initialize to something
> invalid (that is what Juliane suggested)
> 3. throw the old introspection away and always write it as
> |introspection.variables("velocity").component_mask|. Is that what you want?
>
> I don't like 3 because it removes compile time checks to runtime.

3 requires a run-time lookup in a map or list for every access. We do these in 
the innermost loop, so this seems unappealingly slow.


> I don't
> think we gain much by 2 over 1 but I am happy to do it if you guys prefer it.

Why would we want to use ASPECT for a problem where we don't have a velocity? 
If we assume that we always have one, then 1 seems good enough.

My apologies for being slow to catch up with the current state of the patch -- 
haven't gotten much ASPECT reviewing done lately :-(